src/QueryBuilder.jsx

```diff
@@ -142,7 +142,22 @@ export default class QueryBuilder extends React.Component {
             valueEditor: ValueEditor
         };
     }
-
+    
+  
+    componentWillReceiveProps(nextProps) {
+        let schema  = {...this.state.schema};
```

Comment:
Since we are passing the schema down to all the children, I think its time we switch to using the React Context. We can use the regular `contextTypes` and `childContext` props for now and then switch to the Provider/Consumer based context which will be in 16.3.

This will simplify the passing of the schema down the tree. What say?

Reply:
Will need a test as well, as the coverage needs to be kept up :-)

Reply:
okay... looking for a crash course in tests and "regular contextTypes and childContext props "

Reply:
Okay I think I get the regular contextTypes and childContext props ..
the QueryBuilder would get a  childContextTypes: {schema: Proptypes.object} and a getChildContext function to return schema then children require same and  use this.context.schema

Reply:
But now after reading the 16.3  Provider/Consumer based context... seems like an inbetween update would be unnecessary as the new implementation is much different. Looks cool though

Reply:
If we do it with the current `childContextTypes` based approach, the move to the new Provider/Consumer is not too hard and just an API change with a new wrapper Consumer Component around the Rule and RuleGroup components. 

Oh, and you (@hellofantastic) would be one of the first few folks to implement this on a real-world Open source React project!!

Reply:
The fame the glory! Right now half way through figuring out tests.